pytorch_lightning/callbacks/model_checkpoint.py

```diff
@@ -353,12 +360,23 @@ def on_save_checkpoint(
     def on_load_checkpoint(
         self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", callback_state: Dict[str, Any]
     ) -> None:
-        self.best_model_score = callback_state["best_model_score"]
-        self.best_model_path = callback_state["best_model_path"]
-        self.best_k_models = callback_state.get("best_k_models", self.best_k_models)
-        self.kth_best_model_path = callback_state.get("kth_best_model_path", self.kth_best_model_path)
-        self.kth_value = callback_state.get("kth_value", self.kth_value)
+        dirpath_from_ckpt = callback_state.get("dirpath", self.dirpath)
+
+        if self.dirpath == dirpath_from_ckpt:
```

Comment:
I think if dirpath is not set in the next run, a new version is assigned and checkpoints will be added there. This happens during `on_pretrain_routine_start`, thus the final checkpoint is decided after that hook is called. But the checkpoint states are loaded before that, so `self.dirpath` may not be initialized correctly by that time.

`on_pretrain_routine_start/end` is under deprecation. https://github.com/PyTorchLightning/pytorch-lightning/pull/11794/files
the alternative `on_fit_start` for it, will again be a problem for us since this is called after callback states are loaded in a general case.

Reply:
Can't we call `__resolve_ckpt_dir` here in that case?

Reply:
or maybe I was rather thinking of moving the call from `on_pretrain_routine_start` to `setup` which will solve the problem. Need to think if there's any edge case here where it will be an issue.

just trying to avoid calling a private method inside a public hook.